### DISCUSSION TITLE
Customizing the interaction with the user

### DISCUSSION THREAD

**Discussion by margual56** [Byron/google-apis-rs#409] [Q&A] [answered]
Hi! First of all, thank you for this amazing project 💯 

## Context
I'm trying to get the token for the user, with a limited scope.

I am calling the `CalendarHub::new(...).events().[...].doit().await`.

## What I want to do
Currently, after calling `.doit()`, the crate *automatically prints* "Please direct your browser to &lt;URI&gt; and follow the instructions displayed there."

However, I want to be the one telling the user what to do. I want to retrieve the link that the user has to open, and show it how I want.

I know of the `delegate` method, but I don't see how I could use it to do this, because the `begin` method only gives `Method` information, not the actual URI...

## Question (summary)
How can I get the URI for the user to open after calling `.doit().await`?

Thanks!

EDIT: I have just realized that this might me more on the `yup_oauth2` side... I'm sorry if that's the case 😅

**Comment by Byron** [comment, marked as answer]
> EDIT: I have just realized that this might me more on the yup_oauth2 side... I'm sorry if that's the case 😅

Indeed, that's where the message is coming from. There might be more issues around that, like localization, and that would be the place to talk about solutions.

However, it should be possible for you to pre-authenticate requests and then just pass a string (the token) to the authenticator when instantiating the hub. That way you should have much more control over how the interaction is presented. This might still not be possible without rewriting parts of `yup-oauth2`, in which case it's most definitely the right place to get in touch with the maintainers.

**Reply by margual56** [reply]
Thank you for the answer, I have opened an [issue there](https://github.com/dermesser/yup-oauth2/issues/192), let's see what can be done :)
